segments/base_segment.go

```diff
@@ -0,0 +1,700 @@
+	// The PMB number should be reported in the Second Line of Address (e.g., PMB 1234).
+	// As an alternative, the entire address can be reported in the First Line of Address; for example, 5678 Main Street PMB 1234.
+	//
+	// Eliminate internal messages such as: “Do not mail”, “Attorney”, “Charge-off”, “Chapter 13”, “Fraud”, “Trustee”, “Estate of”, “Care of”, “M/R” (Mail Returned), etc.
+	//
+	// Exhibit 13 provides general rules for address reporting.
+	// Do not enter data furnisher's address in this field.
+	FirstLineAddress string `json:"firstLineAddress" validate:"required"`
+
+	// Contains second line of address, if needed, such as apartment or unit number, or private mailbox number (PMB).
+	//
+	// Eliminate internal messages such as: “Do not mail”, “Attorney”, “Charge-off”, “Chapter 13”, “Fraud”, “Trustee”, or “Estate of”, “Care of”, “M/R” (Mail Returned), etc.
+	SecondLineAddress string `json:"secondLineAddress,omitempty"`
+
+	// Contains city name for address of primary consumer.
+	// Truncate rightmost positions if city name is greater than 20 characters or use standard 13-character U.S. Postal Service city abbreviations.
+	City string `json:"city" validate:"required"`
+
+	// Contains the standard U.S. Postal Service state abbreviation for the address of the primary consumer.
+	//
+	// Exhibit 14 provides a list of State Codes.
+	State string `json:"state" validate:"required"`
+
+	// Report the Zip Code of the primary consumer’s address.
+	// Use entire field if reporting 9-digit zip codes.
+	// Otherwise, leftjustify and blank fill.
+	ZipCode string `json:"sipCode" validate:"required"`
+
+	// Contains one of the following values for the address reported in fields 40-44:
+	//
+	//  C  =  Confirmed/Verified address Note: Value ‘C’ enables reporting a confirmed or verified address after receiving an address discrepancy notification from a consumer reporting agency.  Report ‘C’ one time after the address is confirmed.
+	//  Y = Known to be address of primary consumer
+	//  N = Not confirmed address
+	//  M = Military address
+	//  S  =  Secondary Address
+	//  B = Business address — not consumer's residence
+	//  U = Non-deliverable address/Returned mail
+	//  D  = Data reporter’s default address
+	//  P  =  Bill Payer Service  — not consumer’s residence
+	//
+	// If indicator not available or unknown, blank fill.
+	AddressIndicator string `json:"addressIndicator"`
+
+	// Contains the one-character residence code of the address reported in fields 40-44.
+	//
+	//  Values available:
+	//    O = Owns
+	//    R = Rents
+	//
+	// If not available or unknown, blank fill.
+	ResidenceCode string `json:"residenceCode"`
+
+	converter
+	validator
+}
+
+func (s *BaseSegment) Parse(record string) error {
+	if utf8.RuneCountInString(record) != BaseSegmentLength {
+		return ErrSegmentInvalidLength
+	}
+
+	fields := reflect.ValueOf(s).Elem()
```

Comment:
I'm not very familiar with Go's reflection, but I do know it [can panic pretty easily](https://github.com/homedepot/goel/pull/2/files). We'll need to fuzz this heavily before a v1.0 release. If it works I'm fine with using it in a safe way.

Reply:
right, but I think that we can use clean code with reflection.
so I investigated code structure in go reflect package. 
many return style is struct instance rather than pointer. 
And I added some error check code to prevent panic too

Reply:
It'll be okay with fuzzing which we can add later on as we have more test/example files.